codice/ddf
# [2.13.x] DDF-4454 Support for the Dominion Test Framework

#### What does this PR do?
This PR adds support for the dominion test framework to DDF. It defines the initial set of Dominion options which allows one to install the DDF kernel or the DDF distribution. Other test-specific options are added as well.

It fixes profile install command to properly fail the command if the profile cannot be located. This allows testing to detect failures when executing the install command via SSH.

It also modifies pom fix to allow a project to reference a feature installed by its own pom file.

#### Who is reviewing it? 

@peterhuffer 
@kcover 
@AzGoalie 
@lessarderic 

#### Select relevant component teams: 
@codice/test 

#### Ask 2 committers to review/merge the PR and tag them here.
@AzGoalie
@djblue
@clockard


#### How should this be tested?

- Successful PR build should be sufficient
- verify profile:install command when it succeeds and fails

#### Any background context you want to provide?
These changes are required to be done on the release branch of DDF since replication is currently based on that branch.

Here is a sample test class:

```
@DDFCommonOptions.ConfigureVMOptionsForTesting
@DDFCommonOptions.ConfigureDebugging
@DDFCommonOptions.ConfigurePorts
@DDFCommonOptions.ConfigureLogging
@DDFCommonOptions.PolicyFile(
  name = "some.name",
  artifact =
      @MavenUrl(
        groupId = "some.group",
        artifactId = "some.artifact",
        version = MavenUrl.AS_IN_PROJECT,
        type = "policy"
      )
)
@DDFOptions.InstallDistribution(solr = true)
@ServiceAdmin
@Feature.Start(
  repository =
      @MavenUrl(
        groupId = "some.group",
        artifactId = "some.artifact",
        version = MavenUrl.AS_IN_PROJECT,
        type = "xml",
        classifier = "features"
      ),
  name = "some.feature"
)
@RunWith(Dominion.class)
public class ITStart {
  @BeforeClass
  public static void setupClass() {
    System.out.println("*** SETUP CLASS");
  }

  @AfterClass
  public static void cleanupClass() {
    System.out.println("*** AFTER CLASS");
  }

  @Before
  public void setup() {
    System.out.println("*** SETUP");
  }

  @After
  public void cleanup() {
    System.out.println("*** AFTER");
  }

  @Test
  public void test1() throws Exception {
    System.out.println("*** TEST 1");
  }

  @Test
  public void test3() throws Exception {
    System.out.println("*** TEST 3");
  }
}
```
#### What are the relevant tickets?
For Jira:
[DDF-4454](https://codice.atlassian.net/browse/DDF-4454)

For GH Issues:
Fixes: #____

#### Screenshots

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
